### PR TITLE
Add Compiler Frontend

### PR DESCRIPTION
**Pull request by J0**
This PR serves to add the relevant components of the parser which will help us convert the transpiled AST produced by the source tree(string) to our Internal Representation of the program(structs).

Through this PR we aim to convert the function:
`function square(    x   ) { return x    *     x; }` 
into our intermediate representation

**TODO**:
- [ ] Write tests for string -> struct serialization
- [ ] Write string-> struct serialization on a version of the AST from source-jslang which we store in a text file
- [ ] Test that the IR produced by frontend can integrate with backend.
- [ ] Find a way to bring the source-jslang lexer + parser over so that we can use it for comprehensive integration testing.

### CODE REVIEW

File: lib-asttoir/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        map.insert("AwaitExpression", true);
+        map.insert("UnaryExpression", true);
+        map.insert("BinaryExpression", true);
+        map.insert("ConditionalExpression", true);
+        map.insert("YieldExpression", true);
+        map.insert("SequenceExpression", true);
+
+        map
+    };
+}
```
Comment (angelsl) [inline lib-asttoir/src/lib.rs:36]: Hmm, I'm wondering if the performance of a simple match is bad enough that you need to use a HashMap like this.

```rust
fn is_expression(typ: &str) -> bool {
    match typ {
        "ThisExpression" => true,
        "Identifier" => true,
        "Literal" => true,
        "ArrayExpression" => true,
        "ObjectExpression" => true,
        "FunctionExpression" => true,
        "ArrowFunctionExpression" => true,
        "TaggedTemplateExpression" => true,
        "MemberExpression" => true,
        "TaggedTemplateExpression" => true,
        "Super" => true,
        "MetaProperty" => true,
        "NewExpression" => true,
        "TaggedTemplateExpression" => true,
        "CallExpression" => true,
        "UpdateExpression" => true,
        "AwaitExpression" => true,
        "UnaryExpression" => true,
        "BinaryExpression" => true,
        "ConditionalExpression" => true,
        "YieldExpression" => true,
        "SequenceExpression" => true,
        _ => false
    }
}
```

Comment (angelsl) [inline lib-asttoir/src/lib.rs:36]: Hmm, I'm wondering if the performance of a simple match is bad enough that you need to use a HashMap like this.

```rust
fn is_expression(typ: &str) -> bool {
    match typ {
        "ThisExpression" => true,
        "Identifier" => true,
        "Literal" => true,
        "ArrayExpression" => true,
        "ObjectExpression" => true,
        "FunctionExpression" => true,
        "ArrowFunctionExpression" => true,
        "TaggedTemplateExpression" => true,
        "MemberExpression" => true,
        "TaggedTemplateExpression" => true,
        "Super" => true,
        "MetaProperty" => true,
        "NewExpression" => true,
        "TaggedTemplateExpression" => true,
        "CallExpression" => true,
        "UpdateExpression" => true,
        "AwaitExpression" => true,
        "UnaryExpression" => true,
        "BinaryExpression" => true,
        "ConditionalExpression" => true,
        "YieldExpression" => true,
        "SequenceExpression" => true,
        _ => false
    }
}
```

File: lib-asttoir/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        map.insert("ThisExpression", true);
+        map.insert("Identifier", true);
+        map.insert("Literal", true);
+        map.insert("ArrayExpression", true);
+        map.insert("ObjectExpression", true);
+        map.insert("FunctionExpression", true);
+        map.insert("ArrowFunctionExpression", true);
+        map.insert("TaggedTemplateExpression", true);
+        map.insert("MemberExpression", true);
+        map.insert("TaggedTemplateExpression", true);
```
Comment (angelsl) [inline lib-asttoir/src/lib.rs:20]: Duplicate of L18

File: lib-asttoir/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        map.insert("Super", true);
+        map.insert("MetaProperty", true);
+        map.insert("NewExpression", true);
+        map.insert("TaggedTemplateExpression", true);
```
Comment (angelsl) [inline lib-asttoir/src/lib.rs:24]: Duplicate of L18

File: lib-asttoir/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                    kind: ir::ExprKind::VarName {
+                        source: ir::TargetExpr::Local {
+                            localidx: VAR_TO_IDX[var_name],
+                            next: None,
+                        },
+                    },
+                },
+            };
+            func_statements.push(return_statement);
+        }
```
Comment (angelsl) [inline lib-asttoir/src/lib.rs:186]: You could use a `match` on `statement_type` here.

**Comment by J0** [comment]
This PR is outdated, we will use Bernard's frontend instead as it is substantially better. Seems like I have much to learn. Thanks for reviewing though @angelsl !

**Pull request closed without merging**